library/std/src/thread/scoped.rs

```diff
@@ -9,23 +9,24 @@ use crate::sync::Arc;
 /// A scope to spawn scoped threads in.
 ///
 /// See [`scope`] for details.
-pub struct Scope<'env> {
+pub struct Scope<'scope, 'env: 'scope> {
     data: ScopeData,
-    /// Invariance over 'env, to make sure 'env cannot shrink,
+    /// Invariance over 'scope, to make sure 'scope cannot shrink,
     /// which is necessary for soundness.
     ///
     /// Without invariance, this would compile fine but be unsound:
     ///
-    /// ```compile_fail
+    /// ```compile_fail,E0373
     /// #![feature(scoped_threads)]
     ///
     /// std::thread::scope(|s| {
-    ///     s.spawn(|s| {
+    ///     s.spawn(|| {
     ///         let a = String::from("abcd");
-    ///         s.spawn(|_| println!("{:?}", a)); // might run after `a` is dropped
+    ///         s.spawn(|| println!("{:?}", a)); // might run after `a` is dropped
     ///     });
     /// });
     /// ```
+    scope: PhantomData<&'scope mut &'scope ()>,
```

Comment:
Yeah, we probably should, but I'm not sure how yet. I've added it to the tracking issue.

Reply:
FWIW, I have an idea of renaming stuff that _could_ improve stuff, but I was gonna officially suggest it in a follow-up PR, but the gist of it would be:

|      Current      |      Suggested     | Notes |
|:-----------------:|:------------------:|:---:|
| `thread::scope()` | `thread::scoped()` <br/> or <br/> `thread::with_scope()` | debatable nit |
|       `'env`      |      `'scope`      | I don't feel strongly about this one either |
|       `'scope`            |             `'spawner`       | but I do for this one
| `Scope` |   `ScopedSpawner`  <br/> or just <br/> `Spawner` | and thus for this one as well

```rs
fn scoped<'scope, R> (
    // `'spawner` is a higher-order lifetime parameter, thus it may not escape the scope
    // of the closure's body: the spawner is effectively "caged" within this function call.
    f: impl for<'spawner> FnOnce(&'spawner ScopedSpawner<'spawner, 'scope>) -> R
) -> R
```

```rs
impl<'spawner, '_> ScopedSpawner<'spawner, '_> {
    fn spawn<R, ThreadBody>(
        &'spawner self,
        thread_body: ThreadBody,
    ) -> JoinHandle<'spawner, T>
    where
        // ThreadBody represents logic / a computation that produces some `R`.
        ThreadBody : FnOnce() -> R,
        // Such computation may (and will!) be run in another thread...
        ThreadBody : Send,
        // ... and thus it may be used / be running for as long as the `ScopedSpawner` is.
        ThreadBody : 'spawner,
        // The computed value needs to be able to come back to our thread.
        R : Send,
        // And its implicit drop glue may act as a computation of its own, that may thus
        // be run while a `ScopedSpawner` is around.
        R : 'spawner,
```

Funnily enough, by phrasing the `F : 'spawner` bound as the straightforward "it may be used / may be running while the `ScopedSpawner` is alive"[^1], it turns out we don't need to think about the universal quantification of `'scope : 'env` and all those more complicated considerations. We just have a thread that has to be able to be run(ning) / used while the `ScopedSpawner`, the handle that ensures the thread is joined, is itself dropped.

[^1]: which is actually almost exactly what happens –just a bit too conservative, since there is a tiny instant between the moment where all the threads are joined and their non-collected computed values are dropped and the moment the `ScopedSpawner` itself finally dies

Moreover, we can see that:

```rs
impl<'spawner> ScopedSpawner<'spawner, '_> {
    fn spawn(
       self: &'_ Self,
       thread: impl 'spawner + Send + FnOnce()...,
```

is not that different from the more natural:

```rs
impl<'spawner> ScopedSpawner<'spawner> {
    fn spawn(
        self: &'_ Self,
        thread: impl 'spawner + Send + FnOnce()...,
```
<details><summary>Why "natural"</summary>

Indeed, if we stare at the latter, it's actually the pre-leakpocalypse API.

And the bound / logic itself, modulo `Send`, is actually the same as imagining a collection of drop glues / of `defer`s, since joining a thread on drop is not that different from running that thread's logic directly on drop:

```rs
struct MultiDefer<'multi_defer> /* = */ (
    Vec<Box<dyn 'multi_defer+ FnOnce()>>,
);
impl Drop for MultiDefer<'_> { fn drop(&mut self) {
    for f in mem::take(self.0) { f(); }
}}

/// Look, same bounds as the pre-leakpocalypse API (but for Send)
impl<'multi_defer> MultiDefer<'multi_defer> {
    fn add(&mut self, f: impl 'multi_defer + FnOnce())
    {
       self.0.push(Box::new(f))
```
</details>

Two differences:

  - We have to add that extra "upper bound" on `'spawner` _for convenience_, since a post-leakpocalypse API needs a higher-order lifetime, and a higher-order lifetime, by default, not only expresses a local/caged lifetime, but it also covers the range of `'static` or other big lifetimes (our spawner won't be living for that long!). Hence the need to opt-out of that "range of big lifetimes", and use a `for<'spawner where 'spawner : 'scope>` quantification. That `where` clause can't be written there and then in current Rust, so it is written inside `ScopedSpawner`'s definition, which thus needs to carry that otherwise useless extra lifetime parameter.

  - For convenience as well, we say that we want a `&'spawner Self` rather than just a `&'_ Self`, since, if we capture (a borrow of) a spawner, and want that capture to be usable / not to dangle while the spawner is alive, it means we need that borrow to last for `'spawner` as well: `'_ = 'spawner`.
    But I wonder if that one is really necessary / if Rust can't figure that one out already at the call sites 🤔

Reply:
I don't think we should rename Scope to Spawner, because I imagine Scope might be able to do more than just spawning threads in the future. For example, I can imagine something like `scope.is_panicked()` to check if any of the threads panicked. Or `scope.num_running_threads()` to see how many threads are currently still running in this scope. Etc.

Reply:
But let's have that discussion on the tracking issue. ^^